quality: Add test for ".func is" everywhere except the test suite

bmgxyz opened the pull request:
Added a test that checks every line not in the test suite for the string
".func is". This will prevent future code from comparing functions using "is"
rather than isinstance(). Partially fixes #13111.

<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->

bmgxyz committed:
quality: Add test for ".func is" everywhere except the test suite

Added a test that checks every line not in the test suite for the string
".func is". This will prevent future code from comparing functions using "is"
rather than isinstance(). Partially fixes #13111.

ylemkimon commented on sympy/utilities/tests/test_code_quality.py:
```diff
@@ -178,6 +180,8 @@ def test_this_file(fname, test_file):
             if (implicit_test_re.search(line) and
                     not filter(lambda ex: ex in fname, import_exclude)):
                 assert False, message_implicit % (fname, idx + 1)
+            if ".func is" in line and "test" not in fname:
```
There may be whitespace between `.` and `func` and/or between `func` and `is`. Why don't use regex like other checks? Also use `test_file_re` like line 162 above or add `else` to it and check `.func is` there.

bmgxyz replied:
Yeah, regex would probably be a good idea. I didn't know about that whitespace case, though. Can there be an arbitrary number of spaces between any of them?

smichr replied:
Based on limited experience, it seems that the intervening whitespace cannot include a newline so using a regex on line (logical or physical, I'm not sure which is being referenced in the code shown) should work. Would an unlikely, but possibly false positive, occur if the text appeared in a comment or string on the line? e.g. `print("do not use foo.func is bar when testing")`

ylemkimon replied:
From Python documentation:
> Except at the beginning of a logical line or in string literals, the whitespace characters space, tab and formfeed can be used interchangeably to separate tokens.

Of course, it's bad coding style, but `code_quality` test is for checking it.

bmgxyz replied:
Interesting. The regex I used is ```\.\s*func\s+is```. Based on some tests, I think that might work. I guess a string or comment false positive could occur, but I'm not sure how to test for that. Is it likely to happen?

bmgxyz committed:
quality: Update ".func is" test to use regex

Previous code used "in" to check for ".func is" and whether or not the file
is part of the test suite, but now a regex is used to accomplish the same
task. Also, the new regex considers cases where additional whitespace exists
between the ".", "func", and "is".

bmgxyz commented:
I'm not sure why this failed on CI. Everything passed locally, and I'm seeing similar stories on other PRs. What am I missing?

gxyd commented:
Its just an issue with SymPy master branch. Probably nothing do with your changes. See this https://github.com/sympy/sympy/issues/13355#issuecomment-334446269

gxyd commented:
I have restarted the failing jobs. (Only on this PR)

bmgxyz commented:
Are these new failures still a result of that issue you mentioned, or is there something I can/should do to my code? :confused:

gxyd commented:
Shouldn't the tests instead 'fail' (code_quality test)? Since in the current master branch, we do have a lot of cases where '.func is' is currently there. Isn't this what this PR introduces?

gxyd commented:
Ah, never mind (I was thinking of `func = self.func` and then `func is Add` without a `.` but that is a very special case). +1 I am merging this.
